### DISCUSSION TITLE
General excessive nesting rule?

### DISCUSSION THREAD

**Discussion by aaime** [pmd/pmd#3325] [Ideas]
I see that PMD has some specialized rules to avoid deep nesting for if statements.
CheckStyle has a triplet of rules to limit nesting fors, ifs and try.

However, looking around, I'm not finding a general nesting limit rule, e.g. something that would complain if there is, for example, an if nested in a for nested in a lambda nested in a switch nested in a couple of levels of ifs (you get the idea I believe).

If a bit of code has a mix of structures nested, and the nesting becomes too deep, the ability to undrestand the code goes down quickly. In presence of automatic formatting that limits the number of columns, readability quickly suffers too.
E.g., here is an example that just ends up looking bad, and in which code intentions and structure begin to get lost (as automatically formatted by Google Java formatter, with AOSP code style):
https://github.com/geoserver/geoserver/blob/9e40cbbccb9626b8d08516c8eacaf5a8f168189b/src/web/core/src/main/java/org/geoserver/web/GeoServerBasePage.java#L271

What about a rule that flags these cases?

**Comment by oowekyala** [comment]
This could potentially be useful... But to be honest [cognitive complexity](https://pmd.github.io/latest/pmd_rules_java_design.html#cognitivecomplexity) also quantifies nesting. I think it achieves what you want while being more general than just counting nesting level. See also #3295

I think it might need some tests for anonymous classes, like the example you linked.

**Reply by aaime** [reply]
I did look into cognitive complexity already... the issue is that is "also quantifies nesting", but not only.

GeoTools and GeoServer are two large code bases (millions of lines of code), and the QA effort I'm implementing is done fully in scarce and disconnected spare time hours. I'm confident I could work on a cleanup that deals with excessive nesting, cause the code blocks are basically already laid out, just need to be factored in separate methods/classes. The same cannot be said for a generic cognitive complexity violation, that might instead require a rework of the algorithms involved.

**Reply by oowekyala** [reply]
> ...the code blocks are basically already laid out, just need to be factored in separate methods/classes. The same cannot be said for a generic cognitive complexity violation, that might instead require a rework of the algorithms involved.

I think it can though. If you have very nested blocks, you might also have to rework your algorithm. One of the design goals of cognitive complexity, as i understand it, is that you can reduce complexity by factoring out blocks in other methods. I don't think there's anything fundamentally different between the refactorings those metrics ask for.

Regardless, if you really want to implement this other metric I think it could fit within PMD's main distribution. Probably we should investigate the differences between the two rules before publishing the new one, and only enable one of them in the quickstart ruleset. That is, if you'd like the rule to be part of the main codebase anyway.
